### PR TITLE
Bootstrap/use read loop

### PR DESCRIPTION
**Pull request by Ben-PH**
* [x] document all added functions
* [ ] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

**Review by Ben-PH — COMMENTED** [review]
In reference to https://github.com/massalabs/massa/issues/3842#issuecomment-1527681707

also sends message from server in a single message, making better use of the channel. This was done in this PR as the inefficiency was revealed while refining the peek helper method.

**Review by modship — APPROVED** [review]

File: massa-bootstrap/src/bindings.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+mod server;
+use std::{
+    io::{self, ErrorKind},
+    time::{Duration, Instant},
+};
+
+pub(crate) use client::*;
+pub(crate) use server::*;
+
+trait BindingReadExact: io::Read {
```
Comment (Ben-PH) [inline massa-bootstrap/src/bindings.rs:11]: for a side-by-side view of this diff, you can exclude the fmt and module re-arrangement commit.

File: massa-bootstrap/src/bindings/client.rs [hunk 2]
```diff
@@ -75,25 +73,24 @@ impl BootstrapClientBinder {
...
-        let sig_array = peek_buff.as_slice()[0..SIGNATURE_SIZE_BYTES]
+        // read the known-len component of the message
+        let known_len = SIGNATURE_SIZE_BYTES + self.size_field_len;
+        let mut known_len_buff = vec![0u8; known_len];
+        // TODO: handle a partial read
+        self.read_exact_timeout(&mut known_len_buff, deadline)
+            .map_err(|(err, _consumed)| err)?;
+
+        // construct the signature
+        let sig_array = known_len_buff.as_slice()[0..SIGNATURE_SIZE_BYTES]
```
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/client.rs:86]: see conversation here: https://github.com/massalabs/massa/issues/3842 for motivation behind reverting the use of peek. Original intent was to make the very notion of needing to handle a partial read, as far as the stream state is concerned, as obsolete. If this approach _could_ be valuable, using it now is very premature

File: massa-bootstrap/src/bindings/client.rs [hunk 3]
```diff
@@ -169,24 +171,35 @@ impl BootstrapClientBinder {
...
-            // send old previous message
-            self.duplex.write_all(prev_message)?;
+            // Provide the signature saved as the previous message
+            write_buf.extend(prev_message);
         } else {
-            // there was no previous message
-
-            //update current previous message
+            //update previous message
             self.prev_message = Some(Hash::compute_from(&msg_bytes));
```
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/client.rs:178]: `@`me revert me

File: massa-bootstrap/src/bindings/server.rs [hunk 4]
```diff
@@ -197,20 +200,17 @@ impl BootstrapServerBinder {
...
 
-        // send message
-        self.duplex.write_all(&msg_bytes)?;
+        // send the data
+        self.duplex.set_write_timeout(duration)?;
+        self.duplex.write_all(&stream_data)?;
 
-        // save prev sig
+        // update prev sig
         self.prev_message = Some(Hash::compute_from(&sig.to_bytes()));
```
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/server.rs:214]: Should make for more efficient use of the stream

File: massa-bootstrap/src/bindings/server.rs [hunk 5]
```diff
@@ -245,16 +246,16 @@ impl BootstrapServerBinder {
...
             .0
         };
 
-        // read message, and discard the peek
-        let mut msg_bytes = vec![0u8; peek_len + (msg_len as usize)];
+        // read the rest of the message
+        let mut msg_bytes = vec![0u8; msg_len as usize];
         self.duplex.read_exact(&mut msg_bytes)?;
-        let msg_bytes = &msg_bytes[peek_len..];
+        // let msg_bytes = &msg_bytes[known_len..];
```
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/server.rs:258]: `@`me delme

File: massa-protocol-worker/src/handlers/peer_handler/tester.rs [hunk 6]
```diff
@@ -124,7 +124,9 @@ impl InitConnectionHandler for TesterHandshake {
                     //TODO: Check ip we are connected match one of the announced ips
                     {
                         let mut peer_db_write = self.peer_db.write();
-                        peer_db_write.index_by_newest.retain(|_, peer_id_stored| peer_id_stored != &peer_id);
+                        peer_db_write
+                            .index_by_newest
+                            .retain(|_, peer_id_stored| peer_id_stored != &peer_id);
```
Comment (Ben-PH) [inline massa-protocol-worker/src/handlers/peer_handler/tester.rs:129]: only a cargo fmt thing that was triggered by CI

**Comment by Ben-PH** [comment]
Request re-review as many changes were made in response to #3885

File: massa-protocol-worker/src/handlers/peer_handler/mod.rs [hunk 7]
```diff
@@ -431,7 +431,9 @@ impl InitConnectionHandler for MassaHandshake {
                     peer_db_write
                         .index_by_newest
                         .insert(Reverse(announcement.timestamp), peer_id.clone());
-                    peer_db_write.index_by_newest.retain(|_, peer_id_stored| peer_id_stored != peer_id);
+                    peer_db_write
+                        .index_by_newest
+                        .retain(|_, peer_id_stored| peer_id_stored != peer_id);
```
[the remark this answers is not in the document]
Comment (modship) [reply massa-protocol-worker/src/handlers/peer_handler/mod.rs:436]: solved

**Review by modship — APPROVED** [review]

**Pull request merged**